File: 03-LinkedList/CPP/ListNode.cpp
//
//  ListNode.cpp
//  03-LinkedList
//
//  Created by Anzeron on 2021/12/17.
//

#include "ListNode.hpp"

using namespace std;

ListNode* create(int arr[], int n) {

    ListNode* head = new ListNode(arr[0]);
    ListNode* node = head;
    for(int i = 1 ; i < n ; i++){
        node->next = new ListNode(arr[i]);
        node = node->next;
    }

    return head;
}

// 打印以head为头结点的链表信息内容
void print(ListNode* head) {

    ListNode* curNode = head;
    while(curNode != NULL){
        cout << curNode->val << " -> ";
        curNode = curNode->next;
    }

    cout << "NULL" << endl;

    return;
}

// 释放以head为头结点的链表空间
void clean(ListNode* head) {

    ListNode* curNode = head;
    while(curNode != NULL){
        ListNode* delNode = curNode;
        curNode = curNode->next;
        delete delNode;
    }

    return;
}

void testListNode() {
    int arr[] = {1,2,3,5};
    ListNode *h = create(arr,4);
    print(h);
    clean(h);
}

File: 03-LinkedList/Swift/ListNode.swift
//
//  ListNode.swift
//  03-LinkedList
//
//  Created by Anzeron on 2021/12/21.
//

import Foundation

public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init() { self.val = 0; self.next = nil; }
    public init(_ val: Int) { self.val = val; self.next = nil; }
    public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
}

func create(_ arr: [Int]) -> ListNode? {
    let dummyHead = ListNode(-1)
    var last = dummyHead
    
    for num in arr {
        let node = ListNode(num)
        last.next = node
        
        last = node
    }
    
    return dummyHead.next
}

func print(_ head: ListNode?) -> Void {
    
    var node = head
    var text = ""
    while node != nil {
        text = text + "\(node!.val)" + " -> "
        
        node = node?.next
    }
    
    text = text + "nil"
    
    print(text)
}

func testListNode() -> Void {
    let head = create([1,2,3,4,5])
    print(head)
    
}
